README.md
# haskrypt


ChangeLog.md
# Changelog for haskrypt

## Unreleased changes


test/ShaSpec.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}

module ShaSpec
  ( spec,
  )
where

import Data.Char
import Import
import RIO.List (repeat)
import qualified RIO.ByteString as B
import Sha (Digest (..), sha256)
import Test.Hspec

digest_from_string :: [Char] -> Digest
digest_from_string = Digest . B.pack . fmap fromIntegral . go
  where
    go (x1 : x2 : xs) = (16 * char_to_nibble x1 + (char_to_nibble x2)) : go xs
    go [] = mempty
    go _ = error "digest_from_string"
    char_to_nibble c | c >= '0' && c <= '9' = ord c - ord '0'
    char_to_nibble c | c >= 'A' && c <= 'F' = ord c - ord 'A' + 10
    char_to_nibble c | c >= 'a' && c <= 'f' = ord c - ord 'a' + 10
    char_to_nibble _ = error "char to nibble"

spec :: Spec
spec = do
  describe "sha256" $ do
    it "handles empty input" $ do
      sha256 "" `shouldBe` digest_from_string "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
    it "handles small input" $ do
      sha256 "hest" `shouldBe` digest_from_string "b831b3e336ecb131ebc5014393c084bf5d1580854212f64df9ed9226feebc4ae"
    it "handles medium-small input" $ do
      sha256 (mconcat . take 64 $ repeat "h") `shouldBe` digest_from_string "52f9b94822ac1eb7e78d68580e70793f084d490986a9cd40214d3b4375b850b1"
    it "handles medium input" $ do
      sha256 (mconcat . take 1000 $ repeat "hesterfint") `shouldBe` digest_from_string "a1623425022871796310384c675ea19a70a8e8fa40442839d4b7f7d1dda51eea"

    it "handles immense input" $ do
      sha256 (mconcat . take 1000000 $ repeat "hesterfint") `shouldBe` digest_from_string "55698d06b8a00c056314ac5ce0cdbc7481e473bfe78722be569044dcda5345a9"


test/UtilSpec.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}

module UtilSpec
  ( spec,
  )
where

import Data.Char
import Import
import qualified RIO.ByteString as B
import Test.Hspec
import Util (bytestring_to_w32, chunk, sha_padding, w32_to_w8)

breplicate :: Int -> Char -> ByteString
breplicate n = B.replicate n . fromIntegral . ord

spec :: Spec
spec = do
  describe "chunk" $ do
    it "handles empty input" $ chunk 4 "" `shouldBe` ([], Nothing)
    it "handles chunk-sized input" $
      chunk 4 "xxxx" `shouldBe` (["xxxx"], Nothing)
    it "handles many-chunk-sized input" $
      chunk 4 (breplicate 60 'x') `shouldBe` (replicate 15 "xxxx", Nothing)
    it "handles off-sized input" $
      chunk 4 "xxxxy" `shouldBe` (["xxxx"], Just "y")
    it "handles off-sized many-chunk input" $
      chunk 4 "zzzzxxxxyyy" `shouldBe` (["zzzz", "xxxx"], Just "yyy")
  describe "bytestring_to_w32" $ do
    it "extracts Word32 from start of bytestring" $
      bytestring_to_w32 0 (B.pack [1, 2, 3, 4]) `shouldBe` 0x01020304
    it "extracts Word32 from middle of bytestring" $
      bytestring_to_w32 2 (B.pack [1, 2, 3, 4, 5, 6, 7, 8]) `shouldBe` 0x03040506
  describe "w32 to w8" $ do
    it "can convert 0" $ do
      w32_to_w8 0 `shouldBe` [0, 0, 0, 0]
    it "can convert numbers" $ do
      w32_to_w8 0x12345678 `shouldBe` [0x12, 0x34, 0x56, 0x78]
  describe "sha_padding" $ do
    it "pads empty input" $ do
      let res = sha_padding 0
      B.length res `shouldBe` 64
      res `shouldBe`( B.pack $ 0x80 : (replicate 63 0))
    it "pads small input" $ do
      let res = sha_padding 4
      B.length res `shouldBe` 60
      res `shouldBe` (B.pack $ 0x80 : (replicate 51 0) <> [0, 0, 0, 0, 0, 0, 0, 4 * 8])

    it "pads near chunk-sized input" $ do
      let res = sha_padding 62
      B.length res `shouldBe` 128 - 62
      res `shouldBe` (B.pack $ 0x80 : (replicate (128 - 62 - 9) 0) <> [0, 0, 0, 0, 0, 0, 1, 240])


app/Main.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE TemplateHaskell #-}
module Main (main) where

import Import
import Run
import RIO.Process
import Options.Applicative.Simple
import qualified Paths_haskrypt

main :: IO ()
main = do
  (options, ()) <- simpleOptions
    $(simpleVersion Paths_haskrypt.version)
    "Header for command line arguments"
    "Program description, also for command line arguments"
    (Options
       <$> switch ( long "verbose"
                 <> short 'v'
                 <> help "Verbose output?"
                  )
    )
    empty
  lo <- logOptionsHandle stderr (optionsVerbose options)
  pc <- mkDefaultProcessContext
  withLogFunc lo $ \lf ->
    let app = App
          { appLogFunc = lf
          , appProcessContext = pc
          , appOptions = options
          }
     in runRIO app run


src/Util.hs
{-# LANGUAGE MultiWayIf #-}
{-# LANGUAGE NoImplicitPrelude #-}

module Util
  ( bytestring_to_w32,
    chunk,
    sha_padding,
    w32_to_w8,
  )
where

import Control.Arrow
import Data.Bits ((.&.), (.|.), Bits, rotateL, shiftR)
import RIO
import qualified RIO.ByteString as B

chunk :: Int -> ByteString -> ([ByteString], Maybe ByteString)
chunk size bs
  | B.null bs = ([], Nothing)
  | B.length bs < size = ([], Just bs)
  | otherwise =
    let (c, cs) = B.splitAt size bs
     in (c :) `first` chunk size cs

bytestring_to_w32 :: Int -> ByteString -> Word32
bytestring_to_w32 idx bs =
  let bytes = fmap fromIntegral . B.unpack . B.take 4 . B.drop idx $ bs
   in case bytes of
        [b3, b2, b1, b0] ->
          (b0 .|. (b1 `rotateL` 8) .|. (b2 `rotateL` 16) .|. (b3 `rotateL` 24))
        _ -> error "bytestring_to_w32"

byte :: (Integral a, Bits a) => a -> Int -> Word8
byte b n = fromIntegral $ (b `shiftR` n) .&. 255

w32_to_w8 :: Word32 -> [Word8]
w32_to_w8 b = fmap (byte b) [24, 16, 8, 0]

sha_padding :: Word64 -> ByteString
sha_padding numBytes =
  let zeroes = 64 - fromIntegral (numBytes + 8 + 1) `mod` 64
      -- use original message length in **bits**
      size = fmap (byte (numBytes * 8)) [56,48 .. 0]
   in B.pack $ 0x80 : (replicate zeroes 0) ++ size


src/Constants.hs
{-# LANGUAGE NoImplicitPrelude #-}

module Constants
  ( sha_initHash,
    sha_roundConstants,
  )
where

import RIO

sha_initHash :: [Word32]
sha_initHash =
  [ 0x6a09e667,
    0xbb67ae85,
    0x3c6ef372,
    0xa54ff53a,
    0x510e527f,
    0x9b05688c,
    0x1f83d9ab,
    0x5be0cd19
  ]

sha_roundConstants :: [Word32]
sha_roundConstants =
  [ 0x428a2f98,
    0x71374491,
    0xb5c0fbcf,
    0xe9b5dba5,
    0x3956c25b,
    0x59f111f1,
    0x923f82a4,
    0xab1c5ed5,
    0xd807aa98,
    0x12835b01,
    0x243185be,
    0x550c7dc3,
    0x72be5d74,
    0x80deb1fe,
    0x9bdc06a7,
    0xc19bf174,
    0xe49b69c1,
    0xefbe4786,
    0x0fc19dc6,
    0x240ca1cc,
    0x2de92c6f,
    0x4a7484aa,
    0x5cb0a9dc,
    0x76f988da,
    0x983e5152,
    0xa831c66d,
    0xb00327c8,
    0xbf597fc7,
    0xc6e00bf3,
    0xd5a79147,
    0x06ca6351,
    0x14292967,
    0x27b70a85,
    0x2e1b2138,
    0x4d2c6dfc,
    0x53380d13,
    0x650a7354,
    0x766a0abb,
    0x81c2c92e,
    0x92722c85,
    0xa2bfe8a1,
    0xa81a664b,
    0xc24b8b70,
    0xc76c51a3,
    0xd192e819,
    0xd6990624,
    0xf40e3585,
    0x106aa070,
    0x19a4c116,
    0x1e376c08,
    0x2748774c,
    0x34b0bcb5,
    0x391c0cb3,
    0x4ed8aa4a,
    0x5b9cca4f,
    0x682e6ff3,
    0x748f82ee,
    0x78a5636f,
    0x84c87814,
    0x8cc70208,
    0x90befffa,
    0xa4506ceb,
    0xbef9a3f7,
    0xc67178f2
  ]


src/Sha.hs
{-# LANGUAGE ApplicativeDo #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}

module Sha
  ( Digest (..),
    sha256init,
    sha256add,
    sha256finish,
    sha256,
  )
where

import Constants (sha_initHash, sha_roundConstants)
import Data.Bits ((.&.), complement, rotateR, shiftR, xor)
import Data.Char
import Numeric (showHex, showIntAtBase)
import RIO
import qualified RIO.ByteString as B
import RIO.List.Partial ((!!))
import Util (bytestring_to_w32, chunk, sha_padding, w32_to_w8)

newtype Digest
  = Digest ByteString
  deriving (Eq)

hex :: (Integral a, Show a) => a -> String
hex x =
  let res = showHex x ""
   in if length res == 1
        then "0" <> res
        else res

bin :: (Integral a, Show a) => a -> String
bin x =
  let repr = chr . (+ ord ('0'))
      res = replicate 32 '0' <> showIntAtBase 2 repr x ""
   in drop (length res - 32) res

instance Show Digest where
  show (Digest b) = concatMap hex $ B.unpack b

data ShaWork
  = ShaWork
      { numBytes :: !Word64,
        pendingWork :: !ByteString,
        hashState :: ![Word32] -- 8 long
      }
  deriving (Show, Eq)

-- get a fresh sha-state upon which we can add work
sha256init :: ShaWork
sha256init =
  ShaWork {numBytes = 0, pendingWork = B.empty, hashState = sha_initHash}

-- add an arbitrary amount of input to the work
sha256add :: ShaWork -> B.ByteString -> ShaWork
sha256add work input =
  let len = fromIntegral $ B.length input
      nextWork =
        work
          { numBytes = numBytes work + len,
            pendingWork = pendingWork work <> input
          }
   in doWork nextWork

-- process as many 512 bit chunks as are available, and store the
-- remainder for next time (or for finishing)
doWork :: ShaWork -> ShaWork
doWork work =
  let pending = pendingWork work
      state = hashState work
      (todo, remainder) = chunk 64 pending
   in work
        { pendingWork = fromMaybe B.empty remainder,
          hashState = foldl' doChunk state todo
        }

-- process a 512 bit chunk of input, modifying the hash-state in the process
doChunk ::
  [Word32] -> -- hash state, size=8
  B.ByteString -> -- 64 bytes, 512 bits
  [Word32] -- new hash state
doChunk current bs =
  let stuff = foldl' (&) current (go <$> [0 .. 63])
   in zipWith (+) current stuff
  where
    !schedule = createMessageSchedule (flip bytestring_to_w32 bs <$> [0, 4 .. 63])
    go i ![!a, !b, !c, !d, !e, !f, !g, !h] =
      let s1 = (e `rotateR` 6) `xor` (e `rotateR` 11) `xor` (e `rotateR` 25)
          ch = (e .&. f) `xor` ((complement e) .&. g)
          temp1 = h + s1 + ch + (sha_roundConstants !! i) + (schedule !! i)
          s0 = (a `rotateR` 2) `xor` (a `rotateR` 13) `xor` (a `rotateR` 22)
          maj = (a .&. b) `xor` (a .&. c) `xor` (b .&. c)
          temp2 = s0 + maj
       in [temp1 + temp2, a, b, c, d + temp1, e, f, g]
    go _ _ = error "doChunk"

-- extend 128 bits of message into 512 bits
createMessageSchedule :: [Word32] -> [Word32]
createMessageSchedule = reverse  . go 48 . reverse
  where
    go :: Int -> [Word32] -> [Word32]
    go size ws | size == 0 = ws
    go size ws'
      | otherwise =
        let ws = next : ws'
            w15 = ws !! 15
            w7 = ws !! 7
            w16 = ws !! 16
            w2 = ws !! 2
            s0 = (w15 `rotateR` 7) `xor` (w15 `rotateR` 18) `xor` (w15 `shiftR` 3)
            s1 = (w2 `rotateR` 17) `xor` (w2 `rotateR` 19) `xor` (w2 `shiftR` 10)
            !next = w16 + s0 + w7 + s1
         in go (size - 1) ws

-- pad the remaining input into a 512 bit block and process,
-- and return resulting hash-state as a digest
sha256finish :: ShaWork -> Digest
sha256finish  = do
  state <- hashState
  pending <- pendingWork
  padding <- sha_padding . numBytes
  let chunk' = pending <> padding
  return $ Digest . B.pack $ concatMap w32_to_w8 $ doChunk state chunk'

sha256 :: ByteString -> Digest
sha256 = sha256finish . sha256add sha256init
